cpp/include/ucxx/memory_handle.h

```diff
@@ -0,0 +1,159 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+#include <memory>
+#include <string>
+#include <utility>
+
+#include <ucp/api/ucp.h>
+
+#include <ucxx/component.h>
+#include <ucxx/context.h>
+
+namespace ucxx {
+
+class RemoteKey;
+
+/**
+ * @brief Component holding a UCP memory handle.
+ *
+ * The UCP layer provides RMA (Remote Memory Access) to memory handles that it controls
+ * in form of `ucp_mem_h` object, this class encapsulates that object and provides
+ * methods to simplify its handling.
+ */
+class MemoryHandle : public Component {
+ private:
+  ucp_mem_h _handle{};       ///< The UCP handle to the memory allocation.
+  size_t _size{0};           ///< The actual allocation size.
+  uint64_t _baseAddress{0};  ///< The allocation's base address.
+
+  /**
+   * @brief Private constructor of `ucxx::MemoryHandle`.
+   *
+   * This is the internal implementation of `ucxx::MemoryHandle` constructor, made private
+   * not to be called directly. This constructor is made private to ensure all UCXX objects
+   * are shared pointers and the correct lifetime management of each one.
+   *
+   * Instead the user should use one of the following:
+   *
+   * - `ucxx::Context::createMemoryHandle`
+   * - `ucxx::createMemoryHandle()`
+   *
+   * @throws ucxx::Error if either `ucp_mem_map` or `ucp_mem_query` fail.
+   *
+   * @param[in] context parent context where to map memory.
+   * @param[in] size    the minimum size of the memory allocation
+   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
+   *                    new memory region.
+   */
+  MemoryHandle(std::shared_ptr<Context> context, const size_t size, void* buffer);
+
+ public:
+  MemoryHandle()                               = delete;
+  MemoryHandle(const MemoryHandle&)            = delete;
+  MemoryHandle& operator=(MemoryHandle const&) = delete;
+  MemoryHandle(MemoryHandle&& o)               = delete;
+  MemoryHandle& operator=(MemoryHandle&& o)    = delete;
+
+  /**
+   * @brief Constructor for `shared_ptr<ucxx::MemoryHandle>`.
+   *
+   * The constructor for a `shared_ptr<ucxx::MemoryHandle>` object, mapping a memory buffer
+   * with UCP to provide RMA (Remote Memory Access) to.
+   *
+   * The allocation requires a `size` and a `buffer`. The actual size of the allocation may
+   * be larger than requested, and can later be found calling the `getSize()` method. The
+   * `buffer` provided may be either a `nullptr`, in which case UCP will allocate a new
+   * memory region for it, or an already existing allocation, in which case UCP will only
+   * map it for RMA.
```

Comment:
*question*: If a buffer is provided, can it be the case that `getSize()` will return a value other than `size`?

Reply:
No, it's only the case when you're handing over the allocation to UCP (i.e., `buffer=nullptr`), but I see why the confusion, thus I pushed https://github.com/rapidsai/ucxx/pull/190/commits/13233f02e9b54a0555f46771e3e46378ccdfe2b0 in an attempt to clarify it. Please let me know if it's still confusing.

Reply:
Looks great, thanks!